### PR TITLE
Make Describe functions available to local objects

### PR DESCRIPTION
**Pull request by smarterclayton**
Make Describe generally useful for showing data about objects a client already
has, vs. only objects that can be retrieved from the server.

Still sorting out the mechanism - want to make Describe useable against objects
on disk or objects that would be part of a config diff.

### CODE REVIEW

**Comment by thockin** [comment]
@jlowdermilk or @ghodss for review or redirect

**Comment by ghodss** [comment]
Can you explain at a high level the approach taken and how it enables describing objects locally? Is it me or does this seem to make describing more pluggable as well? Is that so non-Kube-native objects can be described?

**Comment by smarterclayton** [comment]
> On Feb 27, 2015, at 12:39 AM, Sam Ghods <notifications@github.com> wrote:
> 
> Can you explain at a high level the approach taken and how it enables describing objects locally?
> 
Ideally, any api object should be describable.   There is value in being able to describe objects that do not exist on the api server, as well as display speculative changes (before and after) where the object in question is opaque to the caller.  Thus, a caller may want to pass in the object and get back a capsule "description" to show to an end user.  This is the "describe me" ability that the default describer exposes.  A third party may wish to add more describers - they can wrap the default describer and pass that in to a caller.

As the current describe functions have shown, there's a reasonable value in having multiple objects pulled together to show to the user.  Those objects may also be relevant  when showing different cases - hence the use of the reflection code to allow matching between "what I have to describe" and "what registered describers there are".  Ideally, this code could become more flexible in the future to handle partial matches (if you have events and a pod, you don't also need to pass in a replication controller list).
> Is it me or does this seem to make describing more pluggable as well?
> 
Technically describe is pluggable today, but this tries to make the use case of having an object (from a directory) available and describing it.  For example, I might want to describe the files in a config dir before applying them to the server.  Right now, it is impossible to display them locally, so this begins to enable that use case.

File: pkg/kubectl/describe.go [hunk 1]
```diff
@@ -409,3 +475,138 @@ func getPodStatusForReplicationController(c client.PodInterface, controller *api
...
+	return "", newErrNoDescriber(append([]reflect.Type{exactType}, types...)...)
+}
+
+// Add adds one or more describer functions to the Describer. The passed function must
+// match the signature:
+//
+//     func(...) (string, error)
+//
+// Any number of arguments may be provided.
+func (d *Describers) Add(fns ...interface{}) error {
```
Comment (bgrant0607) [inline pkg/kubectl/describe.go]: Would it simplify anything to just make the argument a map of kind to describer?

**Comment by bgrant0607** [comment]
Re. describing objects in the client: I don't think we should be round-tripping in the client. We really should think about providing "dry run" in the server. 

I'm fine with this PR for now, however.

LGTM. Rerunning Travis.

File: pkg/kubectl/describe.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/kubectl/describe.go]: We wouldn't be able to pass extra fields (in the future) but that's not a short term concern.  I can simplify it for now - my primary goal was to be able to describe a kind from a file.

**Comment by bgrant0607** [comment]
Sorry, "WIP" in the title caused me to drop this. Is it still WIP? If not, could you please remove WIP from the title, rebase, and ensure tests pass?

Handing off to @nikhiljindal.

**Comment by smarterclayton** [comment]
Will rebase and re title

**Comment by smarterclayton** [comment]
Updated.

**Comment by nikhiljindal** [comment]
LGTM

**Pull request merged**